### DISCUSSION TITLE
Struggling to express `LOWER(expr) IS IN values` (+ many deep issues with expressions)

### DISCUSSION THREAD

**Discussion by Expurple** [SeaQL/sea-query#770] [Q&A] [answered]
I have an `expr: sea_query::Expr` and `values: impl Iterator<Item = String>` and I just can't construct this query.

I expected to be able to write something like `Func::lower(expr).is_in(values)`.

Ok, `FunctionCall` doesn't have `is_in`. I skim through its docs, it's convertible to `SimpleExpr`. Let's convert it:

```rust
use tap::Conv;
Func::lower(expr).conv::<SimpleExpr>().is_in(values)
```

But it doesn't have `is_in` either! [Apparently](https://docs.rs/sea-orm/latest/sea_orm/?search=is_in), only `Expr` and `ColumnTrait` have it. Ok, maybe I need to further convert that `SimpleExpr` to `Expr`. Wait, I can't? What do I do now? I tried to resort to raw SQL, but I can't even find a way to convert my `expr` and `values` to their SQL string representations. I'm blocked on my task and I'm writing this post.

---

I want to note that there's a bigger issue here. In the past, I had many other problems with the whole `Expr` / `SimpleExpr` distinction, where some functions accept the latter but not the former. E.g., the mentioned `Func::lower` doesn't accept `SimpleExpr`, I'm lucky that I have `Expr` as my input. But why? I don't understand why there are two types. It's confusing, one often lacks features of other, and conversions between them are often unergonomic or impossible. Something needs to be done about this. Please, at least document the difference. Currently, `Expr` documentation says that it's a

> Helper to build a `SimpleExpr`.

But this seems unclear and unhelpful. Why is the latter "simple", when the former seems like a lower-level building block? Why is the "hepler" a separate type rather than a function? Why aren't all expressions usable in expression contexts, like the one in the title?

**Comment by xamgore** [comment, marked as answer]
You may apply `Expr::expr` to a value of type `FunctionCall`. But I agree :+1:, the API of custom functions is non-trivial at all.

**Reply by Expurple** [reply]
Thank you, this solves my issue. The resulting code: `Expr::expr(Func::lower(expr)).is_in(values)`.

I'm marking this as the answer, but I really don't feel like the root issue is resolved. These type conversions are a mess. I read a lot of docs today and didn't find that method

**Comment by tyt2y3** [comment]
It would be cool if we can write `Func::lower(expr).is_in(values)`.

But the names `expr` / [`simple_expr`](https://dev.mysql.com/doc/refman/8.0/en/expressions.html) comes from MySQL.

**Reply by Expurple** [reply]
> It would be cool if we can write `Func::lower(expr).is_in(values)`.

Well, this could be solved with `ExprTrait` from #771. I'll mention `FunctionCall` there to not forget.

> But the names `expr` / [`simple_expr`](https://dev.mysql.com/doc/refman/8.0/en/expressions.html) comes from MySQL.

Let's continue the discussion about these types under #771
